File: packages/eulerian-smoke-diff/eulerian_smoke_diff/capture.py
# mypy: ignore-errors
"""Inverse-solution capture with the schema-1.1.0 ``gradient_fields`` key (Stack E / Warp).

The canonical capture is the inverse-problem solution: the recovered initial smoke field plus the
autodiff gradient ``∂Loss/∂u₀`` at the recovered point — a consumer of the WU-A ``gradient_fields``
capture key (schema 1.1.0). Built via ``common_warp.capture.write_frames_capture`` (the Stack-E
convention; the manifest is a raw dict carrying ``gradient_fields``).
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
from common_warp.capture import write_frames_capture

from .forward import SmokeDiffConfig
from .sim import InverseSolution, solve_recovery

__all__ = ["CANONICAL_DESCRIPTOR", "default_capture", "write_inverse_capture"]

CANONICAL_DESCRIPTOR = "eulerian-smoke-diff-recover-u0-16sq-seed42"

_STACK = {"name": "warp-stack-e", "version": "1.13", "build_id": "phase-4-batch-1-smoke-diff"}


def write_inverse_capture(
    solution: InverseSolution,
    out_dir: str | Path,
    *,
    descriptor: str = CANONICAL_DESCRIPTOR,
    start_utc: str = "2026-05-31T00:00:00Z",
) -> Path:
    """Write the inverse-solution capture (recovered field + ``gradient_fields``).

    Returns the manifest ``.json`` path. The payload ``.h5`` carries the recovered initial field
    and the ``dLoss_du0`` gradient; the manifest declares the gradient via the schema-1.1.0
    ``gradient_fields`` key. ``start_utc`` is fixed for a byte-reproducible committed fixture."""
    out = Path(out_dir)
    out.mkdir(parents=True, exist_ok=True)

    grad = np.asarray(solution.grad_fields["dLoss_du0"], dtype=np.float64)
    field = np.asarray(solution.recovered_field, dtype=np.float64)
    n = field.shape[0]

    manifest = {
        "schema_version": "1.1.0",
        "sim": {"name": "eulerian-smoke-diff", "category": "volumetric-grid", "variant": "diff"},
        "stack": dict(_STACK),
        "config": {
            "tier": "reference",
            "dims": [n, n],
            "dtype": "f64",
            "seed": 42,
            "params": {
                "iterations": len(solution.loss_trajectory),
                "final_loss": float(solution.loss_trajectory[-1]),
            },
        },
        "run": {
            "step_count": 1,
            "capture_interval": 1,
            "wall_clock_seconds": 0.0,
            "start_utc": start_utc,
        },
        "payload": {"format": "hdf5", "path": f"{descriptor}.h5", "checksum": "sha256:" + "0" * 64},
        "determinism": {"claimed": "bit-exact-same-hw", "atomic_ops": True, "subgroup_ops": False},
        "gradient_fields": [
            {"name": "dLoss_du0", "shape": list(grad.shape), "dtype": "float64", "wrt": "u0"},
        ],
    }

    frames = [(0, {"smoke_density": field, "dLoss_du0": grad}, {})]
    return write_frames_capture(frames, manifest, out, schema_version="1.1.0")


def default_capture(out_dir: str | Path) -> Path:
    """Solve the canonical inverse problem and write its capture."""
    sol = solve_recovery(SmokeDiffConfig())
    return write_inverse_capture(sol, out_dir)

File: packages/heat-equation/web/src/capture.ts
// heat-equation — web-gate canonical capture (new_canonical, spec-ref § 13.1).
//
// Gate scene: fourier-multi at the 128^2 web-gate tier (alpha 0.02,
// dt 10/128^2 = 0.8x the von Neumann bound, 512 steps, checkpoints every
// 128) — the schrodinger-smoke reduced-tier precedent keeps the capture
// bundle small while the visible demo runs larger grids. The IC is built in
// pure-JS f64 (heat64.mjs, the backend algorithm), cast once to f32, and
// stepped on TWO dedicated 128^2 GPU solvers — the FTCS stencil path and
// the spectral path whose per-mode multipliers come from the COMMITTED
// f64 decay table (public/heat-gate-decay-f64.bin; spec-ref § 5.2/§ 8 —
// never WGSL exp). At each checkpoint both f32 fields are read back and the
// diagnostics are evaluated in f64 (total heat, L2, pinned-mode amplitudes,
// Parseval). verify.py re-runs the f64 reference LIVE and compares
// per-checkpoint per-field max_abs against the [defaults.heat-equation]
// budget.
//
// Determinism witness: the driver runs the whole page twice; the field byte
// streams must match exactly (pure grid stencil + fixed Stockham order, no
// atomics on the gated path).

import type {
  CaptureBundle,
  CaptureStepDescriptor,
} from "../../../../common/common-web/src/capture-export.js";
import { field } from "../../../../common/common-web/src/capture-export.js";
import {
  continuousEigenvalue,
  l2Norm,
  makeCanonicalIc,
  parsevalRelErr,
  sinsinAmplitude,
  totalHeat,
} from "./heat64.mjs";
import { HeatGpu } from "./solver.js";

export const GATE = {
  n: 128,
  alpha: 0.02,
  dt: 10 / (128 * 128), // 0.8 * von Neumann bound; r_x + r_y = 0.4
  steps: 512,
  captureInterval: 128,
  modes: [
    [1, 1],
    [5, 3],
    [2, 7],
  ] as Array<[number, number]>,
  amplitudes: [0.5, 0.25, 0.125],
  offset: 1.0,
  diagModes: [
    [1, 1],
    [5, 3],
  ] as Array<[number, number]>,
  descriptor: "fourier-multi-128sq-alpha0.02-step512-webgate",
};

export async function fetchGateDecayF64(): Promise<Float64Array> {
  const resp = await fetch("./heat-gate-decay-f64.bin");
  if (!resp.ok) throw new Error(`decay table fetch failed: ${resp.status}`);
  return new Float64Array(await resp.arrayBuffer());
}

export async function sha256hex(data: Float32Array): Promise<string> {
  const digest = await crypto.subtle.digest(
    "SHA-256",
    data.buffer.slice(data.byteOffset, data.byteOffset + data.byteLength) as ArrayBuffer,
  );
  return Array.from(new Uint8Array(digest))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

export interface GateRun {
  steps: CaptureStepDescriptor[];
  /** sha-256 over the concatenated checkpoint field bytes (both paths). */
  trajectorySha: string;
  /** raw checkpoint snapshots for the PROVE instruments. */
  ftcsAt: Map<number, Float32Array>;
  specAt: Map<number, Float32Array>;
  /** worst pinned-mode relative error of the f32 spectral path vs the
   * continuous f64 golden, across checkpoints. */
  worstModeRelErr: number;
}

/** One full gate run on two dedicated 128^2 solvers (FTCS + spectral). */
export async function runGateScene(
  device: GPUDevice,
  decayF64: Float64Array,
): Promise<GateRun> {
  const params = {
    alpha: GATE.alpha,
    dt: GATE.dt,
    bcKind: 0 as const,
    wallValue: 0,
    useMaterial: false,
    sourceScale: 0,
  };
  const decayF32 = Float32Array.from(decayF64);
  const gpuF = new HeatGpu(device, GATE.n, params, decayF32);
  const gpuS = new HeatGpu(device, GATE.n, params, decayF32);
  try {
    const ic64 = makeCanonicalIc(GATE.n);
    const ic32 = Float32Array.from(ic64);
    gpuF.uploadField(ic32);
    gpuS.uploadField(ic32);
    gpuF.uploadSource(new Float32Array(GATE.n * GATE.n));
    gpuS.uploadSource(new Float32Array(GATE.n * GATE.n));

    const steps: CaptureStepDescriptor[] = [];
    const ftcsAt = new Map<number, Float32Array>();
    const specAt = new Map<number, Float32Array>();
    const shaChunks: Float32Array[] = [];
    let worstModeRelErr = 0;

    const checkpoint = async (step: number): Promise<void> => {
      const tf = await gpuF.readField();
      const ts = await gpuS.readField();
      ftcsAt.set(step, tf);
      specAt.set(step, ts);
      shaChunks.push(tf, ts);
      const tf64 = Float64Array.from(tf);
      const ts64 = Float64Array.from(ts);
      const diags: Record<string, number> = {
        total_heat_ftcs: totalHeat(tf64, GATE.n),
        total_heat_spec: totalHeat(ts64, GATE.n),
        l2_ftcs: l2Norm(tf64, GATE.n),
        t_min: tf.reduce((a, b) => Math.min(a, b), Infinity),
        t_max: tf.reduce((a, b) => Math.max(a, b), -Infinity),
        parseval_rel_err: parsevalRelErr(ts64, GATE.n),
        sim_time: step * GATE.dt,
      };
      for (let mi = 0; mi < GATE.diagModes.length; mi++) {
        const [m, k] = GATE.diagModes[mi];
        const ampF = sinsinAmplitude(tf64, GATE.n, m, k);
        const ampS = sinsinAmplitude(ts64, GATE.n, m, k);
        diags[`amp_ftcs_${m}_${k}`] = ampF;
        diags[`amp_spec_${m}_${k}`] = ampS;
        const idx = GATE.modes.findIndex(([a, b]) => a === m && b === k);
        const amp0 = GATE.amplitudes[idx];
        const expect =
          amp0 * Math.exp(GATE.alpha * continuousEigenvalue(GATE.n, m, k) * step * GATE.dt);
        if (step > 0) {
          worstModeRelErr = Math.max(worstModeRelErr, Math.abs(ampS - expect) / Math.abs(expect));
        }
      }
      steps.push({
        step,
        state: {
          t_ftcs: field(tf, [GATE.n, GATE.n], "f32"),
          t_spec: field(ts, [GATE.n, GATE.n], "f32"),
        },
        diagnostics: diags,
      });
    };

    await checkpoint(0);
    for (let i = 1; i <= GATE.steps; i++) {
      const enc = device.createCommandEncoder();
      gpuF.encodeFtcsStep(enc);
      gpuS.encodeSpectralStep(enc);
      device.queue.submit([enc.finish()]);
      if (i % GATE.captureInterval === 0) {
        await checkpoint(i);
      }
    }

    const total = new Float32Array(shaChunks.reduce((a, c) => a + c.length, 0));
    let off = 0;
    for (const c of shaChunks) {
      total.set(c, off);
      off += c.length;
    }
    const trajectorySha = await sha256hex(total);
    return { steps, trajectorySha, ftcsAt, specAt, worstModeRelErr };
  } finally {
    gpuF.destroy();
    gpuS.destroy();
  }
}

export function makeBundle(run: GateRun, seed: number, wallSeconds: number): CaptureBundle {
  return {
    manifest: {
      schema_version: "1.0.0",
      sim: {
        name: "heat-equation",
        category: "volumetric-grid",
        variant: "ftcs-plus-spectral-etd1-webgpu",
      },
      stack: { name: "webgpu-f32", version: "0.0.1", build_id: "phase-6-heat-equation-web" },
      config: {
        tier: "test",
        dims: [GATE.n, GATE.n],
        dtype: "f32",
        seed,
        params: {
          alpha: GATE.alpha,
          dt: GATE.dt,
          dx: 1 / GATE.n,
          n: GATE.n,
          safety: 0.8,
          modes: GATE.modes,
          amplitudes: GATE.amplitudes,
          offset: GATE.offset,
        },
      },
      run: {
        step_count: GATE.steps,
        capture_interval: GATE.captureInterval,
        wall_clock_seconds: wallSeconds,
        start_utc: "2026-07-08T00:00:00Z",
      },
      payload: {
        format: "hdf5",
        path: `${GATE.descriptor}.h5`,
        checksum: "sha256:" + "0".repeat(64),
      },
      determinism: { claimed: "bit-exact-same-hw", atomic_ops: false, subgroup_ops: false },
    },
    steps: run.steps,
  };
}
